sys/net/application_layer/nanocoap/cache.c

```diff
@@ -0,0 +1,327 @@
+    }
+
+    return ce;
+}
+
+nanocoap_cache_entry_t *nanocoap_cache_request_lookup(const coap_pkt_t *req)
+{
+    uint8_t cache_key[SHA256_DIGEST_LENGTH];
+    nanocoap_cache_entry_t *ce;
+
+    /* generate cache key */
+    nanocoap_cache_key_generate(req, cache_key);
+
+    /* check if cache key already exists */
+    ce = nanocoap_cache_key_lookup(cache_key);
+
+    return ce;
+}
+
+int nanocoap_cache_process(const uint8_t *cache_key, unsigned request_method,
+                           const coap_pkt_t *resp, size_t resp_len)
+{
+    nanocoap_cache_entry_t *ce;
+    ce = nanocoap_cache_key_lookup(cache_key);
+
+    /* This response is not cacheable. */
+    if (resp->hdr->code == COAP_CODE_CREATED) {
+        /* NO OP */
+    }
+    /* This response is not cacheable. However, a cache MUST mark any
+       stored response for the deleted resource as not fresh.
+    */
+    else if (resp->hdr->code == COAP_CODE_DELETED) {
+        if (ce) {
+            /* set max_age to now(), so that the cache is considered
+             * stale immdiately */
+            ce->max_age = ztimer_now(ZTIMER_SEC);
+        }
+    }
+    /* When a cache that recognizes and processes the ETag response
+       option receives a 2.03 (Valid) response, it MUST update the
+       stored response with the value of the Max-Age Option included
+       in the response (explicitly, or implicitly as a default value;
+       see also Section 5.6.2). For each type of Safe-to-Forward
+       option present in the response, the (possibly empty) set of
+       options of this type that are present in the stored response
+       MUST be replaced with the set of options of this type in the
+       response received. (Unsafe options may trigger similar
+       option-specific processing as defined by the option.)
+    */
+    else if (resp->hdr->code == COAP_CODE_VALID) {
+        if (ce) {
+            /* refresh max_age() */
+            uint32_t max_age = 60;
+            coap_opt_get_uint(resp, COAP_OPT_MAX_AGE, &max_age);
+            ce->max_age = ztimer_now(ZTIMER_SEC) + max_age;
+        }
+        /* TODO: handle the copying of the new options (if changed) */
+    }
+    /* This response is not cacheable. However, a cache MUST mark any
+       stored response for the changed resource as not fresh.
+    */
+    else if (resp->hdr->code == COAP_CODE_CHANGED) {
+        if (ce) {
+            /* set max_age to now(), so that the cache is considered
+             * stale immdiately */
+            ce->max_age = ztimer_now(ZTIMER_SEC);
+        }
+    }
+    /* This response is cacheable: Caches can use the Max-Age Option
+       to determine freshness and (if present) the
+       ETag Option for validation.
+    */
+    else if (resp->hdr->code == COAP_CODE_CONTENT) {
+        uint32_t now = ztimer_now(ZTIMER_SEC);
+        /* cache entry is stale */
+        if (ce && (ce->max_age < now)) {
+            nanocoap_cache_del(ce);
+        }
```

Comment:
If I understand it correctly, when `add_by_key` is able to update the cache entry with a new response, this deletion is not necessary and would make porting ETag support much simpler.